File: src/Authzi.MicrosoftOrleans.Interoperability/Authorization/IncomingGrainCallAuthorizationFilter.cs
﻿using System.Security.Claims;
using System.Threading.Tasks;
using Authzi.Security;
using Authzi.Security.AccessToken;
using Authzi.Security.Authorization;
using IdentityModel;
using Microsoft.Extensions.Logging;
using Orleans;
using Orleans.Runtime;

namespace Authzi.MicrosoftOrleans.Authorization
{
    // ReSharper disable once ClassNeverInstantiated.Global
    public class IncomingGrainCallAuthorizationFilter : GrainAuthorizationFilterBase, IIncomingGrainCallFilter
    {
        public IncomingGrainCallAuthorizationFilter(IAccessTokenVerifier accessTokenVerifier,
            // ReSharper disable once SuggestBaseTypeForParameter
            IAuthorizationExecutor authorizeHandler, ILogger<IncomingGrainCallAuthorizationFilter> logger)
            : base(accessTokenVerifier, authorizeHandler)
        {
            Logger = logger;
        }

        public async Task Invoke(IIncomingGrainCallContext context)
        {
            if (AuthorizationAdmission.IsRequired(context))
            {
                var claims = await AuthorizeAsync(context);
                
                var grainType = context.Grain.GetType();

                if (grainType.BaseType == typeof(GrainWithClaimsPrincipal))
                {
                    var claimsIdentity = new ClaimsIdentity(claims, 
                        "", JwtClaimTypes.Subject, JwtClaimTypes.Role);
                    
                    var claimsPrincipal = new ClaimsPrincipal(claimsIdentity);
                    RequestContext.Set(ConfigurationKeys.ClaimsPrincipalKey, claimsPrincipal);
                }
                
                Log(LoggingEvents.IncomingGrainCallAuthorizationPassed,
                    grainType.Name, context.InterfaceMethod.Name);
            }

            await context.Invoke();
        }
    }
}

File: src/Authzi.MicrosoftOrleans/IncomingGrainCallAuthorizationFilter.fs
namespace Authzi.MicrosoftOrleans.Authorization

open Authzi.MicrosoftOrleans
open Authzi.Security.AccessToken
open Authzi.Security.Authorization
open Authzi.Security
open Microsoft.Extensions.Logging
open Newtonsoft.Json
open Orleans
open Orleans.Runtime
open System.Threading.Tasks

type IncomingGrainCallAuthorizationFilter(accessTokenVerifier: IAccessTokenVerifier,
    authorizeHandler: IAuthorizationExecutor, logger: ILogger<IncomingGrainCallAuthorizationFilter>) as this =
    inherit GrainAuthorizationFilterBase(accessTokenVerifier, authorizeHandler, logger)
    member _.AuthorizeAsync(context) = base.AuthorizeAsync(context)
    member _.Log(eventId, grainTypeName, interfaceMethodName) = base.Log(eventId, grainTypeName, interfaceMethodName)
    interface IIncomingGrainCallFilter with
        member _.Invoke(context: IIncomingGrainCallContext) =
            task {
                if AuthorizationAdmission.IsRequired context then
                    let grainType = context.Grain.GetType()
                    if grainType.BaseType = typeof<GrainWithClaimsPrincipal> then
                        let! claims = this.AuthorizeAsync(context)
                        let serializedClaims = JsonConvert.SerializeObject(claims)
                        RequestContext.Set(ConfigurationKeys.ClaimsPrincipalKey, serializedClaims)
                    
                    this.Log(LoggingEvents.IncomingGrainCallAuthorizationPassed,
                        grainType.Name, context.InterfaceMethod.Name)

                do! context.Invoke()
            } :> Task